mypy/build.py

```diff
@@ -866,34 +866,36 @@ def is_file(path: str) -> bool:
 
 
 USER_SITE_PACKAGES = \
-    '"import site; print(site.getusersitepackages()); print(*site.getsitepackages(), sep=\'\\n\')"'
+    'from __future__ import print_function; import site; print(site.getusersitepackages());' \
+    'print(*site.getsitepackages(), sep="\\n")'
 VIRTUALENV_SITE_PACKAGES = \
-    '"from distutils.sysconfig import get_python_lib; print(get_python_lib())"'
+    'from distutils.sysconfig import get_python_lib; print(get_python_lib())'
 
 
 def call_python(python: str, command: str) -> str:
-    return check_output([python, '-c', command]).decode(sys.stdout.encoding)
+    return subprocess.check_output([python, '-c', command], stderr=subprocess.PIPE).decode('UTF-8')
 
 
-def get_package_dirs(python: Optional[str]) -> List[str]:
+def get_package_dirs(python: str) -> List[str]:
     """Find package directories for given python
 
      This defaults to the Python running mypy."""
-    if package_dirs_cache:
-        return package_dirs_cache
+    if package_dirs_cache.get(python, None):
+        return package_dirs_cache[python]
     package_dirs = []  # type: List[str]
     if python:
         # Use subprocess to get the package directory of given Python
         # executable
-        check = check_output([python, '-V'], stderr=STDOUT).decode('UTF-8')
+        check = subprocess.check_output([python, '-V'], stderr=subprocess.STDOUT).decode('UTF-8')
         assert check.startswith('Python'), \
             "Mypy could not use the Python executable: {}".format(python)
         # If we have a working python executable, query information from it
-        output = call_python(python, USER_SITE_PACKAGES)
-        for line in output.splitlines():
-            if os.path.isdir(line):
-                package_dirs.append(line)
-        if not package_dirs:
+        try:
+            output = call_python(python, USER_SITE_PACKAGES)
+            for line in output.splitlines():
+                if os.path.isdir(line):
+                    package_dirs.append(line)
+        except subprocess.CalledProcessError:
```

Comment:
This `except` should not wrap the `for` loop -- that can be in the `else` clause.